include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -840,6 +840,173 @@ class __static_monotonic_dispatcher<::std::integer_sequence<::std::uint16_t, _X,
+    {
+        _ONEDPL_PRAGMA_UNROLL
+        for (std::uint8_t __i = 0; __i < __vec_size; ++__i)
+            __load_op(__start_idx + __i, __i, __rngs...);
+    }
+
+    template <typename _IdxType, typename _LoadOp, typename... _Rngs>
+    void
+    operator()(/*__is_full*/ std::false_type, _IdxType __start_idx, _LoadOp __load_op, _Rngs&&... __rngs) const
+    {
+        std::uint8_t __elements = std::min(std::size_t{__vec_size}, std::size_t{__full_range_size - __start_idx});
+        for (std::uint8_t __i = 0; __i < __elements; ++__i)
+            __load_op(__start_idx + __i, __i, __rngs...);
+    }
+};
+
+template <typename _TransformOp>
+struct __scalar_store_transform_op
+{
+    _TransformOp __transform;
+    // Unary transformations into an output buffer
+    template <typename _IdxType1, typename _IdxType2, typename _SourceAcc, typename _DestAcc>
+    void
+    operator()(_IdxType1 __idx_source, _IdxType2 __idx_dest, _SourceAcc __source_acc, _DestAcc __dest_acc) const
+    {
+        __transform(__source_acc[__idx_source], __dest_acc[__idx_dest]);
+    }
+    // Binary transformations into an output buffer
+    template <typename _IdxType1, typename _IdxType2, typename _Source1Acc, typename _Source2Acc, typename _DestAcc>
+    void
+    operator()(_IdxType1 __idx_source, _IdxType2 __idx_dest, _Source1Acc __source1_acc, _Source2Acc __source2_acc,
+               _DestAcc __dest_acc) const
+    {
+        __transform(__source1_acc[__idx_source], __source2_acc[__idx_source], __dest_acc[__idx_dest]);
+    }
+};
+
+// TODO: Consider unifying the implementations of __vector_walk, __vector_load, __vector_store, and potentially
+// __strided_loop with some common, generic utility
+template <std::uint8_t __vec_size>
+struct __vector_walk
+{
+    static_assert(__vec_size <= 4, "Only vector sizes of 4 or less are supported");
+    std::size_t __full_range_size;
+
+    template <typename _IdxType, typename _WalkFunction, typename... _Rngs>
+    void
+    operator()(std::true_type, _IdxType __idx, _WalkFunction __f, _Rngs&&... __rngs) const
+    {
+        _ONEDPL_PRAGMA_UNROLL
+        for (std::uint8_t __i = 0; __i < __vec_size; ++__i)
+        {
+            __f(__rngs[__idx + __i]...);
+        }
+    }
+    // For a non-full vector path, process it sequentially. This will always be the last sub or work group
+    // if it does not evenly divide into input
+    template <typename _IdxType, typename _WalkFunction, typename... _Rngs>
+    void
+    operator()(std::false_type, _IdxType __idx, _WalkFunction __f, _Rngs&&... __rngs) const
+    {
+        std::uint8_t __elements = std::min(std::size_t{__vec_size}, std::size_t{__full_range_size - __idx});
+        for (std::uint8_t __i = 0; __i < __elements; ++__i)
+        {
+            __f(__rngs[__idx + __i]...);
+        }
+    }
+};
+
+template <std::uint8_t __vec_size>
+struct __vector_store
```

Comment:
What's a difference from struct `__vector_load`? It looks identic for common body - __store_op/load_op.

Reply:
https://github.com/uxlfoundation/oneDPL/pull/1976/files#r1927807323
In a previous comment, I suggest consolidating these, and we've marked it as a TODO.